File: 프로그래머스/2/12973. 짝지어 제거하기/짝지어 제거하기.py
# 짝지어 제거하기(스택 사용!!)

def solution(s):
    answer = 0
    stack = []

    for i in range(len(s)):
        # 만약 stack에 아무것도 없으면 stack에 넣어준다.
        if len(stack) == 0: 
            stack.append(s[i])
            
        # stack의 마지막이랑 s[i] 랑 같으면 stack 빼준다.
        elif stack[-1] == s[i]: 
            stack.pop()
            
        else:
            stack.append(s[i])

    if len(stack) == 0:
        return 1
    else:
        return 0 

File: 프로그래머스/2/12973. 짝지어 제거하기/짝지어 제거하기.java
import java.util.ArrayDeque;

class Solution {
    public int solution(String s) {
        // 같은 알파벳이 2개 붙어 있는 짝을 찾고 제거하는 문제

        // 문자들을 저장할 스택을 생성 (연결된 동일 문자를 처리하기 위함)
        ArrayDeque<Character> stack = new ArrayDeque<>();

        // 문자열의 모든 문자를 순서대로 확인
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i); // 현재 위치의 문자를 가져옴

            // 스택이 비어있지 않고, 스택의 맨 위에 있는 문자와 현재 문자가 같으면
            if (!stack.isEmpty() && stack.peek() == c) {
                stack.pop(); // 스택에서 맨 위의 문자를 제거 (짝을 이루는 두 문자를 제거)
            } else {
                stack.push(c); // 짝이 없으면 현재 문자를 스택에 추가
            }
        }

        // 스택이 비어 있으면 모든 짝이 제거되었으므로 1을 반환, 그렇지 않으면 0 반환
        return stack.isEmpty() ? 1 : 0;
    }
}